### PR TITLE
AT-61-update-student

### PR DESCRIPTION
**Pull request by bferenczy0**
Added implementation so that the api handles a request to change data in the student database, with error catching in the case of a blank input and associated tests, and returns the updated student data.
Jira: https://dfacademy.atlassian.net/jira/software/projects/AT/boards/4?issueParent=10137&selectedIssue=AT-61

### CODE REVIEW

File: cypress/integration/students.spec.js [hunk 1]
```diff
@@ -19,5 +19,38 @@ describe("Students", function() {
         });
       });
   });
-});
-  
+  it("updates student data", function() {
```
Comment (dearshrewdwit) [inline cypress/integration/students.spec.js:22]: Let's rename this block of tests.

Group them in order by 
1. endpoint - (students, individual student), maybe into separate files
2. HTTP method - (GET, POST, PUT etc)
3. Scenario - (success, failure)

Use  `#describe()` to group tests.

Use `#it()` to assert what should be returned `

File: routes/students.js [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
-    }
-  });
+router.get("/:id", async function (req,res) {
+  const student = await Student.findOne({where: {id: req.params.id}});
+  res.json({student:student});
+});
 
-  res.json({student: student});
+router.put("/:id", async function (req,res) {
+  if (req.body[Object.keys(req.body)] === ""){
```
Comment (dearshrewdwit) [inline routes/students.js:18]: Let's start a bit more simply, and then spot opportunities for refactors
NB: Each of the following needs a corresponding test
> 1. return error if student doesn't exist
> 2. return error if req params don't exist
> 3. return error if update not successful (eg updating a text field over the character length limit)
> 4. return success if success

```js
// pseudocode
const permittedParams = ['firstName', 'lastName']
router.put("/:id", async function (req,res) {
  const student = await Student.findOne({where: {id:req.params.id}})
  if (student === null) {
    res.json({
      status: "failure",
      student: {},
      errors: "Missing record"
  } else if (!Object.keys(req.body).every(key => permittedParams.includes(key)) {
    res.json({
      status: "failure",
      student: {},
      errors: "Invalid params"
    }
  } else {
    student.update(req.body)
      .then(student => {
        res.json({
          status: "success",
          student: updatedStudent,
          errors: []
        })
     .catch(error => {
        res.json({
          status: "failure",
          student: {},
          errors: [error.message] // something like this?
        })
    }
  }
}
```

Comment (dearshrewdwit) [reply routes/students.js:18]: We might then be able to clean this up afterwards

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Nice so far

File: cypress/integration/students.spec.js [hunk 3]
```diff
@@ -20,4 +20,64 @@ describe("Students", function() {
...
+    cy.request({
+      method: "PUT",
+      url: "api/students/0"
+    }).then(res => {
+      expect(res.body.status).to.eq("failure");
+      expect(res.body.errors).to.eq("Missing record");
+    });
+  });
+  it("returns error for invalid parameters", function() {
+    cy.request({
```
Comment (dearshrewdwit) [inline cypress/integration/students.spec.js:39]: Do you have access to a hardcoded student id from the createStudent task?

File: cypress/integration/students.spec.js [hunk 4]
```diff
@@ -20,4 +20,64 @@ describe("Students", function() {
...
+        expect(res.body.status).to.eq("failure");
+        expect(res.body.errors).to.eq("No input");
+      });});
+  });
+  it("returns a successful update", function() {
+    cy.request({
+      method: "GET",
+      url: "api/students"
+    }).then(res => {
+      expect(res.body.students[0].id).to.be.a("number");
```
Comment (dearshrewdwit) [inline cypress/integration/students.spec.js:73]: Can we add an assertion on the name so we can see that it changes on a successful update?

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]
Nice - thanks for covering the different cases in tests.

See comments

File: cypress/integration/students.spec.js (hunk 3 above)
Comment (dearshrewdwit) [reply cypress/integration/students.spec.js:39]: I checked - you don't. What about hardcoding student IDs in the cypress task, so you can use them in the test?

There's a tradeoff here - we're testing implicitly the `/api/students` endpoint, as well as the individual student endpoint - but your test is decoupled from knowing about specific student IDs.

In this case I would err on the side of only testing one endpoint, unless I can't get around it.

Comment (bferenczy0) [reply cypress/integration/students.spec.js:39]: Did not realise you could hardcode an id in a cypress task, that does make things a lot easier

File: cypress/integration/students.spec.js (hunk 4 above)
Comment (bferenczy0) [reply cypress/integration/students.spec.js:73]: ``
      cy.request({
      method: "PUT",
      url: "api/students/1",
      body: {firstName: "Fred"}
    }).then(res => {
      expect(res.body.status).to.eq("success");
      expect(res.body.student.firstName).to.eq("Fred");
```
Such as this or do you mean another test afterwards that gets api/students and tests for the new name?

Comment (bferenczy0) [reply cypress/integration/students.spec.js:73]:
```suggestion
      cy.request({
      method: "PUT",
      url: "api/students/1",
      body: {firstName: "Fred"}
    }).then(res => {
      expect(res.body.status).to.eq("success");
      expect(res.body.student.firstName).to.eq("Fred");
```

Such as this or do you mean another test afterwards that gets api/students and tests for the new name?

Comment (dearshrewdwit) [reply cypress/integration/students.spec.js:73]: tldr - it doesn't matter anymore if you're making 1 put request.

If you're no longer making 2 requests - the first request to get all the students, it doesn't matter as much - we can rely on looking at the task for the previous name to verify the update actually happens. If you were getting data first before updating, this extra assertion would have added confidence that the update is working as expected (decreasing reliance on how the students endpoint works)

File: routes/students.js [hunk 5]
```diff
@@ -20,4 +22,49 @@ router.get("/:id", async function (req, res) {
...
+router.get("/:id", async function (req,res) {
+  const student = await Student.findOne({where: {id: req.params.id}});
+  res.json({student:student});
+});
+
+const permittedParams = ["firstName", "lastName", "githubUsername", "email"];
+router.put("/:id", async function (req,res) {
+  const student = await Student.findOne({where:{id:req.params.id}});
+  if (student === null) {
+    res.json({
```
Comment (dearshrewdwit) [inline routes/students.js:34]: can you extract the response object?
```js
const response = {
  status: "failure",
  student: {},
  errors: ""
}
if (student === null) {
  response.error = "missing record"
  res.json(response);
}
```

**Review by RyanVanDijck — APPROVED** [review]

**Pull request merged**